### PR TITLE
New bridge method (fetchContactsByPhone) to obtain the contacts info given an phone number…

### PR DESCRIPTION
**Pull request by vcollantes**

### CODE REVIEW

File: src/post-message.ts [hunk 1]
```diff
@@ -135,6 +135,20 @@ export type IncomingMessageMap = {
...
+    MATCH_CONTACT_DATA: {
+        id: string;
+        type: 'MATCH_CONTACT_DATA';
+        payload: {
+            matches: {
+                [key: string]: {
+                    name?: string;
+                    middleName?: string;
+                    surname?: string;
+                    encodedAvatar?: string;
```
Comment (atabel) [inline src/post-message.ts]: all attributes are optional?

File: src/webview-bridge-cjs.js.flow [hunk 2]
```diff
@@ -57,6 +57,19 @@ declare export function requestContact(
...
+    numbers: Array<string>,
+): Promise<{|
+    matches: {
+        [key: string]: {
+            name?: string,
+            middleName?: string,
+            surname?: string,
+            encodedAvatar?: string,
+        },
+    },
```
Comment (atabel) [inline src/webview-bridge-cjs.js.flow]: Use explicit exact or inexact object types, don't be implicit
```js
matches: {
    [key: string]: {|
        name?: string,
        middleName?: string,
        surname?: string,
        encodedAvatar?: string,
    |},
    ...
}
```

File: src/webview-bridge-cjs.js.flow [hunk 3]
```diff
@@ -65,7 +78,9 @@ declare export function requestSimImsi(): Promise<string | null>;
 
 declare export function requestDeviceImei(): Promise<string | null>;
 
-declare export function requestVibration(type: 'error' | 'success'): Promise<null>
+declare export function requestVibration(
+    type: 'error' | 'success',
+): Promise<null>;
```
Comment (atabel) [inline src/webview-bridge-cjs.js.flow]: why this change? prettier?

File: src/post-message.ts (hunk 1 above)
Comment (vcollantes) [reply src/post-message.ts]: Yes, as I talked to @jmpuerta, all those attributes can come or not

File: src/webview-bridge-cjs.js.flow (hunk 2 above)
Comment (vcollantes) [reply src/webview-bridge-cjs.js.flow]: ok!

File: src/webview-bridge-cjs.js.flow (hunk 3 above)
Comment (vcollantes) [reply src/webview-bridge-cjs.js.flow]: Yes, prettier

File: README.md [hunk 4]
```diff
@@ -76,6 +76,7 @@ Alternatively, you can import the library directly from a CDN:
 -   [checkPermissionStatus](#checkPermissionStatus)
 -   [internalNavigation](#internalNavigation)
 -   [dismiss](#dismiss)
+-   [matchContactData](#matchContactData)
```
Comment (pladaria) [inline README.md]: to me this is a bad name, match contact data? Why *match*? why *singular* contact if you can return many? data? everything is data. 

You are fetching contacts by phone number so:
requestContactsByPhone
fetchContactsByPhone
getContactsByPhone

This way, if in the future you want to fetch by email or name, you can follow this same convention

File: README.md [hunk 5]
```diff
@@ -558,6 +559,23 @@ import {requestVibration} from '@tef-novum/webview-bridge';
 requestVibration('error');
 ```
 
+### matchContactData
+
+Returns contacts info given an array of phone numbers.
+
+```javascript
+matchContactData: (numbers: Array<string>) => Promise<{
+    matches: {
```
Comment (pladaria) [inline README.md]: instead of `matches` I'd call this `contacts`

File: README.md (hunk 4 above)
Comment (vcollantes) [reply README.md]: I gave this name, because of the type of the bridge, that's why I choose that name, but you're  right and in this case I prefer requestContactsByPhone. @atabel what do you think?

File: README.md [hunk 6]
```diff
@@ -558,6 +559,23 @@ import {requestVibration} from '@tef-novum/webview-bridge';
...
+        [key: string]: {
```
Comment (pladaria) [inline README.md]: what's this key? the phone number, an identifier? give it a name

Anyway, do we really need a map here? this could be just a collection

File: README.md (hunk 5 above)
Comment (vcollantes) [reply README.md]: This structure goes in a NPI team TS, please if you don't have access to this TS, please talk to me and I'll share with you. I prefer to not copy a TS in a public repo

File: README.md (hunk 6 above)
Comment (vcollantes) [reply README.md]: this key is the normalized number, maybe yo prefer normalizedNumber instead of key.
As I've written in the prev comment, this is a TS done by NPI team

Comment (pladaria) [reply README.md]: I think it would be better to include it in the collection

File: README.md (hunk 5 above)
Comment (pladaria) [reply README.md]: having a key makes sense when you want to pick items by key, but in this case the key is unknown, so useless.

If this method returns a collection, use a collection (array of contacts)

File: README.md (hunk 6 above)
Comment (pladaria) [reply README.md]: having a key makes sense when you want to pick items by key but in this case the key is unknown to the caller so it is useless

if you want to return a collection, use a collection (array of contacts). Keep it simple

Comment (vcollantes) [reply README.md]: Unkown? this key is necessary to the caller and identify uniquely the information.

File: README.md (hunk 4 above)
Comment (jmpg93) [reply README.md]: That's a great suggestion. I like fetchContactsByPhone. I will update the TS if everyone agrees.

Comment (vcollantes) [reply README.md]: fetchContactsByPhone is the name function in the bridge, this is not the type MATCH_CONTACT_DATA, but If you prefer you can change the type.

File: README.md (hunk 5 above)
Comment (jmpg93) [reply README.md]: That makes sense. Let's remove the extra nesting.

Comment (vcollantes) [reply README.md]: Great, thanks @jmpg93

File: README.md (hunk 6 above)
Comment (jmpg93) [reply README.md]: The map is keyed by the numbers you request on the bridge. The map is nice to fastly check if the number you request has a match and ensures one match per number.

Comment (pladaria) [reply README.md]: Ok, so that key is not the formatted number as vcollantes said

Comment (vcollantes) [reply README.md]: Yes, as I talked to @carlosfernandez and @jpmg93, a number can have multiple contacts,for this reason, instead of a map, the returned structure will be an array, and the phoneNumber will be inside.

Comment (pladaria) [reply README.md]: To confirm, is the key the number used to perform the request or a normalized number?

Comment (vcollantes) [reply README.md]: The key was the normalizedNumber, please check the old structure (because now will be an array):
"payload":{ 
         "34600000001":{ 
            "firstName":"Josh",
            "middleName":"Claus",
            "lastName":"Snow",
            "encodedAvatar":"RXJlcyB1biBoYWNrZXI="
         },
         "34600000002":{ 
            "firstName":"Josh",
            "middleName":"Claus",
            "lastName":"Snow",
            "encodedAvatar":"RXJlcyB1biBoYWNrZXI="
         },
         "34600000003":{ 
            "firstName":"Josh",
            "middleName":"Claus",
            "lastName":"Snow",
            "encodedAvatar":"RXJlcyB1biBoYWNrZXI="
         },
         "34600000004":{ 
            "firstName":"Josh",
            "middleName":"Claus",
            "lastName":"Snow",
            "encodedAvatar":"RXJlcyB1biBoYWNrZXI="
         }
   }

where "34600000001" is the normalized number

Comment (pladaria) [reply README.md]: then, if the key is the normalized number, it cannot be used to check that the requested number has a match, as @jmpg93 said. That key should be the request number

Comment (vcollantes) [reply README.md]: The requested number is the normalizedNumber, when I request the numbers, previosly I normalize those numbers, check the petition (it's an array of normalized numbers)

"type":"FETCH_CONTACT_DATA",
   "id":"5",
   "payload":{ 
      "numbers":[ 
         "34600000001",
         "34600000002",
         "34600000003",
         "34600000004"
      ]
   }

Comment (pladaria) [reply README.md]: So you are forcing the caller to implement a number normalization logic (which is not an easy task and probably requires libphonenumber)

I think it would be better to accept any kind of number (local ones and normalized) and leave the native side to perform the matching.

Keep in mind that this bridge can potentially be used by any OB, not only us

Comment (jmpg93) [reply README.md]: That's a good point, but right now we only have one use case. In that use cases, the caller already has the normalized number, so it does not make sense to call numbering services again. 

If new use cases arise in the future, we can iterate over it.

Comment (vcollantes) [reply README.md]: And the normalized number is served by backend, and backend already is using libphonenumber library, so there are no extra problems in this case.

Comment (wyunreal) [reply README.md]: Thats true, at least in account, all normalization number are provided by server, avoiding including libphonenumber in client side, so, I would go with the current approach of having the keys as normalized numbers unless you think there will be another client of this bridge that wont have server providing these nombers

Comment (wyunreal) [reply README.md]: Anyways, this code is a passthrough (this lib) from client code to native code and viceversa, if native part should normalize/denormalize numbers in the future, this code wont change.

File: README.md (hunk 4 above)
Comment (vcollantes) [reply README.md]: Changes done!
